### DISCUSSION TITLE
Using GitVersion locally with GitVersion.Core!

### DISCUSSION THREAD

**Discussion by MrHinsh** [GitTools/GitVersion#3758] [Q&A]
I am trying to get the current version of the GitRepository that my code is sitting in during the runtime of the application. 

While most people use the latest version of my application from Winget, some folks run it in debug locally. This makes it difficult to ascertain what version of the code they are using, especially when they post segments of the log...

I added the "GitVersion.MSBuild" which does give the "next" version, but it fails if folks download the code without cloning the repo.

```
        private static string GetVersionTextForLog()
        {
            Version runningVersion = DetectVersionService2.GetRunningVersion();
            string textVersion = $"v" + ((runningVersion.Major > 1) ? runningVersion : "Local");
            if (textVersion == "vLocal")
            {
                textVersion += " " + GitVersionInformation.MajorMinorPatch;
                
            }
           return textVersion;
        }
```
https://github.com/nkdAgility/azure-devops-migration-tools/blob/7fd7ba9d22a0f61148416637a6e441191f0c75f4/src/MigrationTools.Host/MigrationToolHost.cs#L151

The following occurs if I remove the `.git` folder to simulate this.

```
Severity	Code	Description	Project	File	Line	Suppression State
Error	MSB3073	The command ""C:\Users\...\.nuget\packages\gitversion.msbuild\5.12.0\tools\net48/gitversion.exe" "C:\Users\...\Source\repos\azure-devops-migration-tools-test\src\MigrationTools.Host"  -output file -outputfile obj\gitversion.json" exited with code 1.	MigrationTools.Host	C:\Users\...\.nuget\packages\gitversion.msbuild\5.12.0\tools\GitVersion.MsBuild.targets	9	
```

I would like to be instead able to get the "CurrentVersion" from the repository with code (using `GitVersion.Core`) so that I can handle the exception when the folder is not a git repo.

The documentation for how to use this is very sparse, and I have been trawling through the tests to find an entry point that I can use to make this call.

This is as far as I have been able to get, but I cant seem to get the last centimetre...

```
using System;
using System.Collections.Generic;
using System.Text;
using GitVersion;
using GitVersion.Extensions;
using Microsoft.Extensions.DependencyInjection;
using GitVersion.Helpers;

namespace MigrationTools.Host.Services
{
    public class DetectGitVersion : IDetectGitVersionService
    {
        public const string MainBranch = "master";

        protected static IServiceProvider ConfigureServices(Action<IServiceCollection>? overrideServices = null)
        {
            var services = new ServiceCollection().AddModule(new GitVersionCoreModule());

            overrideServices?.Invoke(services);

            return services.BuildServiceProvider();
        }

        protected static IServiceProvider BuildServiceProvider(IGitRepository repository,
            string? targetBranch = null, IReadOnlyDictionary<object, object?>? configuration = null)
        {
            var options = Options.Create(new GitVersionOptions
            {
                WorkingDirectory = repository.Path,
                ConfigInfo = { OverrideConfig = configuration },
                RepositoryInfo = { TargetBranch = targetBranch }
            });

            var sp = ConfigureServices(services =>
            {
                services.AddSingleton(options);
                services.AddSingleton(repository);
            });
            return sp;
        }
        public void GetGitVariables()
        {
                var options = Options.Create(new GitVersionOptions { WorkingDirectory = workDirectory, Settings = { NoFetch = true } });

                var sp = ConfigureServices(services => services.AddSingleton(options));

                var gitVersionCalculator = sp.GetRequiredService<IGitVersionCalculateTool>();

                gitVersionCalculator.CalculateVersionVariables();
        }
    }
}
```

Any hints would be much appreciated and I do understand that this would not be supported.

**Comment by HHobeck** [comment]
Have you seen the IntegrationTests? The integration tests are executed via API. If I delete the .git folder I'm getting the following error:

![image](https://github.com/GitTools/GitVersion/assets/56404113/f2044cb7-51d0-4130-b798-63e910c6f98c)

Edit: I have test it on main branch which reflects the beta.3 version.

**Reply by MrHinsh** [reply]
Is it posible to have the tool not error out?

**Reply by HHobeck** [reply]
This is something I cannot answer. At this moment it is not implemented this way obviously. Is it a problem for you to catch the DirectoryNotFoundException?

**Reply by MrHinsh** [reply]
It errors out ar compile time locally in Visual Studio. I don't see a way to handle this in MSBuild.

```
Severity	Code	Description	Project	File	Line	Suppression State
Error	MSB3073	The command ""C:\Users\...\.nuget\packages\gitversion.msbuild\5.12.0\tools\net48/gitversion.exe" "C:\Users\...\Source\repos\azure-devops-migration-tools-test\src\MigrationTools.Host"  -output file -outputfile obj\gitversion.json" exited with code 1.	MigrationTools.Host	C:\Users\...\.nuget\packages\gitversion.msbuild\5.12.0\tools\GitVersion.MsBuild.targets	9	
```

**Reply by arturcic** [reply]
I think you installed the GitVersion.Msbuild on a project that is not versioned, am I right?

**Reply by MrHinsh** [reply]
No.

The project is versioned, but sometimes folks download a zipped copy of the code.

I don't want to ever have a compile error if the .git folder is missing. A runtime exception when I try to get the version is fine.

**Comment by MrHinsh** [comment]
Why was this closed when unanswered?

**Comment by arturcic** [comment]
I think the best option for you would be to have 
` <DisableGitVersionTask>true</DisableGitVersionTask>` set at the project level

https://github.com/GitTools/GitVersion/blob/main/src/GitVersion.MsBuild/msbuild/tools/GitVersion.MsBuild.props#L28

This will allow to build the project from the zip.

Then in your CI/CD you just set programatically the 
```xml
dotnet build some.csproj -p:DisableGitVersionTask=false
```

And it should work
